ruby/ql/lib/codeql/ruby/frameworks/ActionController.qll

```diff
@@ -85,6 +86,16 @@ class ActionControllerActionMethod extends Method, HTTP::Server::RequestHandler:
   // not end at an explicit render or redirect
   /** Gets the controller class containing this method. */
   ActionControllerControllerClass getControllerClass() { result = controllerClass }
+
+  /**
+   * Gets a route to this handler, if one exists.
+   * May return multiple results.
+   */
+  ActionDispatch::Route getARoute() {
+    result.getController() + "_controller" =
+      ActionDispatch::underscore(namespaceDeclaration(controllerClass)) and
+    this.getName() = result.getAction()
```

Comment:
This looks like a candidate for a bad join, but I don't know if joining on `getName()` and `getAction()` alone can yield a large set in practive. It should be possible to join simultaneously on name and controller, see e.g. https://github.com/github/codeql/pull/7738 for an example.

Reply:
I'm still getting my head around this problem and its mitigation. I've replaced this line with a call to `isActionControllerMethod(this, result.getAction(), controllerClass)` (https://github.com/github/codeql/pull/7061/commits/749dc092ae37e542582a03e84974a255fc7fb7bf), which will hopefully force a join simultaneously on `getName()`, `getAction()` and `controllerClass`. Is that enough to protect against this issue or do we need to do something else?

Reply:
If you add a new predicate
```ql
pragma[nomagic]
private predicate isRoute(
  ActionDispatch::Route r, string name, ActionControllerControllerClass controllerClass
) {
  r.getController() + "_controller" =
    ActionDispatch::underscore(namespaceDeclaration(controllerClass)) and
  name = r.getAction()
}
```

and change the body to

```ql
    exists(string name |
      isRoute(result, name, controllerClass) and
      isActionControllerMethod(this, name, controllerClass)
    )
```

then simultaneous joining on `name` and `controllerClass` should be ensured.